File: lib/feed_forward.dart
import 'gpu_tensor.dart';
import 'nn.dart';

class FeedForward extends Module {
  final Layer w1; // Up-projection (embedSize -> embedSize * 4)
  final Layer w2; // Down-projection (embedSize * 4 -> embedSize)

  FeedForward(int dim)
    : w1 = Layer(dim, dim * 4, useGelu: true),
      w2 = Layer(dim * 4, dim, useGelu: false);

  Tensor forward(Tensor x, List<Tensor> tracker) {
    // 1. Expand and Activate (GELU)
    final hidden = w1.forward(x, tracker);
    // 2. Contract back to original dimension
    final out = w2.forward(hidden, tracker);
    return out;
  }

  @override
  List<Tensor> parameters() => [...w1.parameters(), ...w2.parameters()];
}

File: python/dart_cuda/core/layers/feed_forward.py
"""FeedForward MLP block — mirrors `lib/core/layers/feed_forward.dart`."""

from __future__ import annotations

from typing import List

from ..tensor.gpu_tensor import Tensor
from .nn import Layer, Module


class FeedForward(Module):
    def __init__(self, dim: int):
        self.w1 = Layer(dim, dim * 4, useGelu=True)
        self.w2 = Layer(dim * 4, dim, useGelu=False)

    def forward(self, x: Tensor, tracker: List[Tensor]) -> Tensor:
        hidden = self.w1.forward(x, tracker)
        return self.w2.forward(hidden, tracker)

    def parameters(self) -> List[Tensor]:
        return [*self.w1.parameters(), *self.w2.parameters()]
